###  Import Data and Required Packages
####  Importing Pandas, Numpy, Matplotlib, Seaborn and Warings Library.

In [ ]:
import pandas as pd

In [7]:
# Load the dataset
df = pd.read_csv("../data/raw/Hotel_train.csv")



#### Show Top 5 Records

In [10]:
df.head()

Reservation-id Gender  Age         Ethnicity Educational_Level  \
0        39428300      M   40            Latino       High-School   
1        77491756      F   49  African American       High-School   
2        73747291      F   42         caucasian           College   
3        67301739      M   25            Latino       High-School   
4        77222321      F   62  African American              Grad   

        Income Country_region      Hotel_Type Expected_checkin  \
0         <25K          North      City Hotel       2015-07-01   
1  50K -- 100K           East  Airport Hotels       2015-07-01   
2         <25K           East  Airport Hotels       2015-07-02   
3        >100K          South      City Hotel       2015-07-02   
4    25K --50K           East  Airport Hotels       2015-07-03   

  Expected_checkout  ... Meal_Type  Visted_Previously  Previous_Cancellations  \
0        2015-07-02  ...        BB                 No                      No   
1        2015-07-02  ...        BB                 No                      No   
2        2015-07-06  ...        BB                 No                      No   
3        2015-07-03  ...        BB                 No                      No   
4        2015-07-04  ...        BB                 No                      No   

     Deposit_type Booking_channel Required_Car_Parking Reservation_Status  \
0      No Deposit          Online                  Yes          check-out   
1  Non-Refundable          Online                  Yes          check-out   
2      Refundable          Online                  Yes          check-out   
3      No Deposit           Agent                  Yes          check-out   
4      Refundable          Direct                   No          check-out   

  Use_Promotion Discount_Rate   Room_Rate  
0           Yes             0  106.440241  
1            No             0  158.039564  
2            No             0  212.610780  
3           Yes             5  106.146407  
4           Yes            10  159.384439  

[5 rows x 24 columns]

#### Shape of the dataset

In [15]:
df.shape

(27499, 24)

###  Check Missing values

In [20]:
df.isna().sum()

Reservation-id               0
Gender                       0
Age                          0
Ethnicity                    0
Educational_Level            0
Income                       0
Country_region               0
Hotel_Type                   0
Expected_checkin             0
Expected_checkout            0
Booking_date                 0
Adults                       0
Children                     0
Babies                    4755
Meal_Type                    0
Visted_Previously            0
Previous_Cancellations       0
Deposit_type                 0
Booking_channel              0
Required_Car_Parking         0
Reservation_Status           0
Use_Promotion                0
Discount_Rate                0
Room_Rate                    0
dtype: int64

#### hadle missing valus

In [26]:
df['Babies'] = df['Babies'].fillna(0)  # Assume no babies if missing
df.isna().sum()


Reservation-id            0
Gender                    0
Age                       0
Ethnicity                 0
Educational_Level         0
Income                    0
Country_region            0
Hotel_Type                0
Expected_checkin          0
Expected_checkout         0
Booking_date              0
Adults                    0
Children                  0
Babies                    0
Meal_Type                 0
Visted_Previously         0
Previous_Cancellations    0
Deposit_type              0
Booking_channel           0
Required_Car_Parking      0
Reservation_Status        0
Use_Promotion             0
Discount_Rate             0
Room_Rate                 0
dtype: int64

###  Check Duplicates

In [29]:
df.duplicated().sum()

np.int64(0)

#### There are no duplicates  values in the data set

###  Check data types

In [33]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27499 entries, 0 to 27498
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Reservation-id          27499 non-null  int64  
 1   Gender                  27499 non-null  object 
 2   Age                     27499 non-null  int64  
 3   Ethnicity               27499 non-null  object 
 4   Educational_Level       27499 non-null  object 
 5   Income                  27499 non-null  object 
 6   Country_region          27499 non-null  object 
 7   Hotel_Type              27499 non-null  object 
 8   Expected_checkin        27499 non-null  object 
 9   Expected_checkout       27499 non-null  object 
 10  Booking_date            27499 non-null  object 
 11  Adults                  27499 non-null  int64  
 12  Children                27499 non-null  int64  
 13  Babies                  27499 non-null  float64
 14  Meal_Type               27499 non-null

###  Checking the number of unique values of each column

In [36]:
df.nunique()

Reservation-id            27495
Gender                        2
Age                          53
Ethnicity                     4
Educational_Level             4
Income                        4
Country_region                4
Hotel_Type                    3
Expected_checkin            901
Expected_checkout           929
Booking_date               1133
Adults                        5
Children                      3
Babies                        3
Meal_Type                     3
Visted_Previously             2
Previous_Cancellations        2
Deposit_type                  3
Booking_channel               3
Required_Car_Parking          2
Reservation_Status            3
Use_Promotion                 2
Discount_Rate                 7
Room_Rate                 21087
dtype: int64

###  Check statistics of data set

In [41]:
df.describe()

Reservation-id           Age        Adults      Children        Babies  \
count    2.749900e+04  27499.000000  27499.000000  27499.000000  27499.000000   
mean     5.016461e+07     43.977454      2.232118      1.743882      0.309466   
std      2.886911e+07     15.303148      1.180172      0.721972      0.579941   
min      3.154000e+03     18.000000      1.000000      1.000000      0.000000   
25%      2.523943e+07     31.000000      1.000000      1.000000      0.000000   
50%      5.014996e+07     44.000000      2.000000      2.000000      0.000000   
75%      7.518288e+07     57.000000      3.000000      2.000000      0.000000   
max      9.999900e+07     70.000000      5.000000      3.000000      2.000000   

       Discount_Rate     Room_Rate  
count   27499.000000  27499.000000  
mean        8.293029    155.645134  
std         7.841725     38.314829  
min         0.000000    100.000000  
25%         0.000000    124.000000  
50%         5.000000    149.453032  
75%        10.000000    181.000000  
max        30.000000    250.000000

# 📊 Dataset Insights base on statistics

## **1. General Information**
- The dataset contains **27,499 reservations**.
- The `Reservation-id` ranges from **3,154 to 99,999,000**, suggesting unique identifiers.

## **2. Age Distribution**
- **Average guest age:** **44 years** (± 15.3 years).
- **Age range:** **18 to 70 years**.
- **Interquartile Range (IQR):** **31 to 57 years** → Most guests are middle-aged.

## **3. Group Composition (Adults, Children, Babies)**
- **Adults per reservation:** 
  - Average: **~2.23** 
  - Max: **5**
- **Children per reservation:**
  - Average: **~1.74**
  - Max: **3**
- **Babies per reservation:**
  - Average: **~0.31**
  - Max: **2**
- **Insight:** At least **75% of bookings have no babies**.

## **4. Discount Rate**
- **Average discount:** **8.29%** (± 7.84%).
- **50% of bookings (median) have a discount of 5% or less**.
- **25% of bookings have no discount at all**.
- **Maximum discount:** **30%** (high promotional offers).

## **5. Room Rates**
- **Average room rate:** **$155.65** (± $38.31).
- **Price range:** **$100 to $250**.
- **Interquartile Range (IQR):** **$124 to $181**.

## **6. Potential Business Insights**
- **🎯 Age Targeting:** Most guests are between **31 and 57**, so marketing efforts should focus on this age group.
- **👨‍👩‍👧‍👦 Group Size Optimization:** Most bookings are for **small families (2 adults, 1-2 children)** → Consider family packages.
- **💰 Discount Strategy:** Since **many bookings receive no discount**, targeted promotions could increase bookings.
- **🏨 Room Pricing:** With a **high standard deviation in room rates**, implementing tiered pricing based on room type may be effective.


